runtime/compiler/runtime/J9Profiler.cpp

```diff
@@ -1730,6 +1730,7 @@ TR_BlockFrequencyInfo::getFrequencyInfo(
       while (bciCheck.getCallerIndex() > -1)
          {
          TR_InlinedCallSite *callSite = &comp->getInlinedCallSite(bciCheck.getCallerIndex());
+         bciCheck.setByteCodeIndex(callSite->_byteCodeInfo.getByteCodeIndex());
```

Comment:
I'm not sure I understand the need for this part (and similar further below in `getFrequencyInfo()`). And something about it doesn't seem quite right. I think this makes `bciCheck` contain a bytecode index that isn't within the inlined site specified by its inlined site index, but is rather within the caller, e.g.
```
Call Stack Info
CalleeIndex CallerIndex ByteCodeIndex CalleeMethod
       0         -1          25       Foo.foo()     [site 0 is foo() as called from -1:25]
       1          0          37       Foo.bar()     [site 1 is bar() as called from  0:37]
       2          1          14       Foo.baz()     [site 2 is baz() as called from  1:14]
```
If we start with BCI `2:0`, i.e. the very beginning of the bytecode for inlined site 2 (`baz()`), then I think this creates the following list of pairs:
```
(<site 2>, 2:14)
(<site 1>, 1:37)
(<site 0>, 0:25)
```
If I'm right, then these are very strange BCI values. And the BCI values that would have been in the pairs before your change already should have included bytecode indices:
```
(<site 2>, 2:0)
(<site 1>, 1:14)
(<site 0>, 0:37)
```
So maybe now I can guess what the purpose of this line is: Is it to get the bytecode index of the outermost call, i.e. the call from the root method? If so, then I would suggest one of the following instead:
- Specify the entire BCI of the call from the caller: `(<2>, 1:14), (<1>, 0:37), (<0>, -1:25)`;
- Specify only the bytecode index (not BCI): `(<2>, 14), (<1>, 37), (<0>, 25)`; or
- Add an entry for the root method: `(<2>, 2:0), (<1>, 1:14), (<0>, 0:37), (null, -1:25)`.

I thought the first would be easiest, and it might still be the easiest thing to do as far as just this loop is concerned, since it would only involve deleting this `setByteCodeIndex()` and moving the assignment to `bciCheck` to before `push_back(make_pair(...))`, but in light of the 'grafting point' search below, I think probably the last option is the least disruptive

Reply:
Thanks a lot @jdmpapin the review and suggestion. I see your point, even though `ComputeEffectiveCallerIndex` does not use the full bytecode info at all, I see having odd bytecode info as you pointed out is not desirable.

So to understand, with 3rd solution, do you mean we match methods in the call stack and in the end, verify the bytecode index of the root method? 

In that case, if the root method calls some other method that contains multiple same call sites, we would still see the same issue we are seeing without this fix right? 
For example, if I make changes to your example, 

```
Call Stack Info
CalleeIndex CallerIndex ByteCodeIndex CalleeMethod
       0         -1          25       Foo.foo()     [site 0 is foo() as called from -1:25]
       1          0          37       Foo.bar()     [site 1 is bar() as called from  0:37]
       2          1          14       Foo.baz()     [site 2 is baz() as called from  1:14]
       3          0          41       Foo.bar()     [site 3 is bar() as called from  0:41]
       4          3          14       Foo.baz()     [site 4 is baz() as called from  3:14]
```

In above case, if we query for `4:0`, when it goes through the call stack, list would look like this. 
`(<4>, 4:0), (<3>, 3:14), (<0>, 0:41), (null, -1:25)`. We would find methods at call stack `4 <- 3 <- 0 <- -1` and `2 <- 1 <- 0 <- -1` same, and computeEffectiveCallerIndex would return 2 for the callsite `4:0`. 


Unless I missed something or misunderstood solution 3, I like solution 2. We can document the intended use of `computeEffectiveCallerIndex` using comments. What do you think?

Reply:
First, sorry - I should have written something like `<site n method>` rather than `<site n>`, since those are `TR_OpaqueMethodBlock*`. So now `<n>` can refer to the result of `getInlinedCallSiteMethod()` for site `n`.

> So to understand, with 3rd solution, do you mean we match methods in the call stack and in the end, verify the bytecode index of the root method?

You'd still have to check _all_ of the bytecode indices, like this PR already does in `computeEffectiveCallerIndex()`. It's just that each of those indices would correspond in the expected way to the inlined site index that it's bundled with. I guess that `computeEffectiveCallerIndex()` would need to compare against the bytecode index of the _next_ element though.

The problem with solutions 1 and 2 is that the "looking for a grafting point" path maintains a single `callStack`, which is both provided to `computeEffectiveCallerIndex()` and used in a way that looks at the inlined site index of the BCI. And (assuming that path works...) it requires the inlined site index that is currently already specified by the BCI. To use one of the first suggestions, this `callStack` would need to be replaced by two slightly different data structures that need to be kept in sync, or else we'd have to create a (slightly modified) copy each time we call `computeEffectiveCallerIndex()`, and I think that would be error prone

Reply:
Thanks a lot for the clarification, It makes sense. Also having `NULL` entry in the call Stack does not impact the grafting point search path and seems to be functionally fine.  I have made the changes in the last couple of force pushed. I am going to run the testing builds again for sanity.

Reply:
Oh, I'm realizing now that I suggested null because I was thinking that the first component of the pair was a `TR_InlinedCallSite*`, but since it's really `TR_OpaqueMethodBlock*`, we might as well put the actual root method there